[FLINK-4456] Replace ActorGateway in Task and RuntimeEnvironment

tillrohrmann opened the pull request:
Replaces the `ActorGateway` in `Task` and `RuntimeEnvironment` by interfaces to decouple these components from the actors.

- Introduces a `TaskExecutionStateListener` interface for `TaskExecutionState` update messages
- Replaces the job manager `ActorGateway` by `InputSplitProvider` and `CheckpointNotifier`
- Replaces the task manager `ActorGateway` by `TaskManagerConnection`

The implementations using the `ActorGateways` are

- `InputSplitProvider` --> `TaskInputSplitProvider`
- `TaskExecutionStateListener` --> `ActorGatewayTaskExecutionStateListener`
- `CheckpointNotifier` --> `ActorGatewayCheckpointNotifier`
- `TaskManagerConnection` --> `ActorGatewayTaskManagerConnection`

StephanEwen commented:
I think this is good.
Few comments inline on the names, the hardest problem usually ;-)

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/CheckpointNotifier.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Notifier for checkpoint acknowledge and decline messages in the {@link Task}.
+ */
+public interface CheckpointNotifier {
```
`CheckpointNotifier` is already used for user functions, as an interface to get notifications on completed checkpoints.

How about calling this `CheckpointResponder` or something like that instead?

tillrohrmann replied:
Oh I've overlooked this class. Will rename it into `CheckpointResponder`.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/tasks/InputSplitProvider.java:
```diff
@@ -35,4 +35,16 @@
+	 *
+	 * @param userCodeClassLoader User code class loader to use by the input split provider
+	 */
+	void start(ClassLoader userCodeClassLoader);
```
Is it possible to have the `start()` and `stop()` methods not in the base interface, but only in the Akka specific implementation?

tillrohrmann replied:
The `start` method's intention is to pass in the `userCodeClassLoader` which is currently created in `Task#run` method. There are two other ways to solve the problem. Either creating the user code class loader outside of `Task` where the `InputSplitProvider` is created or to pass the user code class loader via the `getNextInputSplit` method call to the input split provider.

For the first approach: Creating the user code class loader is a blocking operation, so this would have to executed in a future and upon completion we could create the `Task` instance in the `TaskManager`.


For the second approach: We would have to touch more code but I think everywhere were the `getNextInputSplit` method is called, we have access to the user code class loader.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskExecutionStateListener.java:
```diff
@@ -0,0 +1,29 @@
+	 *
+	 * @param taskExecutionState describing the task execution state change
+	 */
+	void notifyTaskExecutionState(TaskExecutionState taskExecutionState);
```
`notifyTaskExecutionStateChanged(...)`?

tillrohrmann replied:
Yes, can rename.

tillrohrmann commented:
I've addressed the comments. In order to solve the problem with the user code class loader, I decided to pass it into the `InputSplitProvider` via the `getNextInputSplit` method. That way, we don't have to know the user code class loader when creating the `InputSplitProvider` implementation.

tillrohrmann commented:
I will rebase the PR on the latest master and if Travis gives green light, I'd like to merge it.
